File: panel/src/main/java/com/effective/android/panel/utils/cutShort/ViVoCutShort.java
package com.effective.android.panel.utils.cutShort;

import android.content.Context;
import android.os.Build;
import android.text.TextUtils;
import android.view.View;

import com.effective.android.panel.utils.DisplayUtil;
import com.effective.android.panel.utils.PanelUtil;

import java.lang.reflect.Method;

/**
 * vivo刘海计算
 * https://dev.vivo.com.cn/documentCenter/doc/103
 * Created by yummyLau on 20-03-27
 * Email: yummyl.lau@gmail.com
 * blog: yummylau.com
 */
public class ViVoCutShort implements DeviceCutShort {

    public static final String VENDOR = "ViVo";

    @Override
    public boolean hasCutShort(Context context) {
        try {
            String manufacturer = Build.MANUFACTURER;
            boolean isVivo = false;
            boolean hasDisplayCutout = false;
            if (!TextUtils.isEmpty(manufacturer) && manufacturer.equalsIgnoreCase("vivo")) {
                isVivo = true;
            }

            ClassLoader cl = context.getClassLoader();
            Class FtFeature = cl.loadClass("android.util.FtFeature");
            Method get = FtFeature.getMethod("isFeatureSupport", Integer.TYPE);
            hasDisplayCutout = (Boolean)get.invoke(FtFeature, 32);
            return isVivo && hasDisplayCutout;
        } catch (Exception var7) {
            var7.printStackTrace();
            return false;
        }
    }

    @Override
    public boolean isCusShortVisible(Context context) {
        return true;
    }

    @Override
    public int getCurrentCutShortHeight(View view) {
        Context context = view.getContext();
        if(!isCusShortVisible(context)){
            return 0;
        }
        return DisplayUtil.dip2px(view.getContext(),27.0F);
    }

}

File: panel-kotlin/src/main/java/com/effective/android/panel/utils/cutShort/ViVoCutShort.kt
package com.effective.android.panel.utils.cutShort

import android.content.Context
import android.os.Build
import android.text.TextUtils
import android.view.View
import com.effective.android.panel.utils.DisplayUtil

/**
 * vivo刘海计算
 * https://dev.vivo.com.cn/documentCenter/doc/103
 * Created by yummyLau on 20-03-27
 * Email: yummyl.lau@gmail.com
 * blog: yummylau.com
 */
class ViVoCutShort : DeviceCutShort {

    override fun hasCutShort(context: Context): Boolean {
        return try {
            val manufacturer = Build.MANUFACTURER
            var isVivo = false
            var hasDisplayCutout = false
            if (!TextUtils.isEmpty(manufacturer) && manufacturer.equals("vivo", ignoreCase = true)) {
                isVivo = true
            }
            val cl = context.classLoader
            val ftFeature = cl.loadClass("android.util.FtFeature")
            val get = ftFeature.getMethod("isFeatureSupport", Integer.TYPE)
            hasDisplayCutout = get.invoke(ftFeature, 32) as Boolean
            isVivo && hasDisplayCutout
        } catch (var7: Exception) {
            var7.printStackTrace()
            false
        }
    }

    override fun isCusShortVisible(context: Context): Boolean {
        return true
    }

    override fun getCurrentCutShortHeight(view: View): Int {
        val context = view.context
        return if (!isCusShortVisible(context)) {
            0
        } else DisplayUtil.dip2px(view.context, 27.0f)
    }

    companion object {
        const val VENDOR = "ViVo"
    }
}